Fix Task Decorator to Work With and Without Feature Flag (AAP-41775)

art-tapin opened the pull request:
##### SUMMARY
Implemented a solution that enables AWX's task system to work with both the legacy dispatcher and the new dispatcherd library based on the `FEATURE_NEW_DISPATCHER` feature flag. Created a registry pattern instead of trying to modify method objects directly.

##### ISSUE TYPE
- Bug Fix

##### COMPONENT NAME
- Other (Dispatcher System)

##### AWX VERSION

##### ADDITIONAL INFORMATION

**Problem:**  
When initially tried to modify methods directly with `cluster_node_heartbeat.apply_async._new_method = ...`, I encountered `AttributeError: 'method' object has no attribute '_new_method'` because Python bound methods don't support attribute assignment.

With the feature flag disabled, I also encountered this scheduler error (probably known):
```
File "/awx_devel/awx/main/dispatch/periodic.py", line 26, in __init__
  self.interval = int(data['schedule'].total_seconds())
AttributeError: 'float' object has no attribute 'total_seconds'
```

**Solution:**  
1. In `publish.py`:
   - Added `ALTERNATIVE_TASK_IMPLEMENTATIONS` registry at module level
   - Modified `apply_async` to check the registry when the feature flag is enabled

2. In `system.py`:
   - Refactored `cluster_node_heartbeat` into two implementations:
     - Original AWX version with `bind_kwargs`
     - New dispatcherd version using `bind=True`
   - Extracted common code into helper functions
   - Added `_get_active_task_ids_from_dispatcherd()` to retrieve running tasks from the new dispatcher
   - Registered the alternative implementation in the central registry

3. Other minor adjustments to ensure compatibility with both systems

**Testing Steps:**
```bash
# (For ansible org) Pull the right image
docker image pull ghcr.io/ansible/awx_devel:feature_dispatcher

# Start AWX with feature dispatcher
COMPOSE_TAG=feature_dispatcher make docker-compose

# In AWX container, test with flag enabled:
# First, enable it in awx/settings/development_defaults.py
awx-manage shell -c "from awx.main.tasks.system import cluster_node_heartbeat; cluster_node_heartbeat.delay()"
# From the logs
grep "Using dispatcherd implementation"

# Test with flag disabled:
# First, enable it in awx/settings/development_defaults.py
awx-manage shell -c "from awx.main.tasks.system import cluster_node_heartbeat; cluster_node_heartbeat.delay()"
```

For more extensive in-the-code debugging logs, checkout the revision (commit): `fix(dispatcher): Implement registry pattern for dispatcher feature flag compatibility`

art-tapin committed:
fix: clean old dispatcher leftovers in settings

art-tapin committed:
refactor(system): extract common heartbeat helpers and split cluster_node_heartbeat

Extract common heartbeat logic into helper functions:  _heartbeat_instance_management: consolidates instance management, health checks, and lost-instance detection.  _heartbeat_check_versions: compares instance versions and initiates shutdown when necessary.  _heartbeat_handle_lost_instances: reaps jobs and marks lost instances offline.

Refactor the original cluster_node_heartbeat to use these helpers and retain legacy behavior (using bind_kwargs).

Introduce adispatch_cluster_node_heartbeat for dispatcherd: uses the control API to retrieve running tasks and reaps them.

Link the two implementations by attaching adispatch_cluster_node_heartbeat as the _new_method on cluster_node_heartbeat.

art-tapin committed:
feat(publish): delegate heartbeat task submission to new dispatcherd implementation

Update apply_async to check at runtime if FEATURE_NEW_DISPATCHER is enabled.

When the task is cluster_node_heartbeat and a _new_method is attached, delegate the task submission to the new dispatcherd implementation.

Preserve the original behavior for all other tasks and fallback on error.

art-tapin committed:
refactor(system): extract task ID retrieval from dispatcherd into helper function

Improves readability of adispatch_cluster_node_heartbeat by extracting
the complex UUID parsing logic into a dedicated helper function.
Adds clearer error handling and follows established code patterns.

art-tapin committed:
fix(dispatcher): Enable task decorator to work with and without feature flag

Implemented a new approach for handling task execution with feature flags
by attaching alternative implementations to apply_async._new_method. This
allows cluster_node_heartbeat to work correctly with both the legacy and
new dispatcher systems without modifying core decorator logic.

AAP-41775

art-tapin committed:
fix(dispatcher): Improve error handling and logging in feature flag implementation

- Add error handling when attaching alternative dispatcher implementation
- Fix method self-reference in apply_async to properly use cls.apply_async
- Document limitations of this targeted approach for specific tasks
- Add logging for better debugging of dispatcher selection
- Ensure decorator timing by keeping method attachment after function definitions

This completes the robust implementation for switching between dispatcher
implementations based on feature flags.

AAP-41775

art-tapin committed:
fix(dispatcher): Implement registry pattern for dispatcher feature flag compatibility

Replaces direct method attribute assignment with a global registry for
alternative implementations. The original approach tried to attach new
methods directly to apply_async bound methods, which fails because bound
methods don't support attribute assignment in Python.

The registry pattern:
- Creates a global ALTERNATIVE_TASK_IMPLEMENTATIONS dict in publish.py
- Registers alternative implementations by task name
- Modifies apply_async to check the registry when feature flag is enabled
- Adds extensive logging throughout the process for debugging

This enables cluster_node_heartbeat to work correctly with both the legacy
and new dispatcher implementations based on the FEATURE_NEW_DISPATCHER flag.

AAP-41775

art-tapin committed:
refactor(dispatcher): Remove excessive logging from dispatcher implementation

Reduces verbose debugging logs while maintaining essential logging for critical
operations. Preserves:
- Task implementation selection based on feature flag
- Registration success/failure messages
- Critical error reporting

Removed:
- Registry content debugging messages
- Repetitive task diagnostics
- Non-essential information logging

AAP-41775

AlanCoding commented:
> AttributeError: 'float' object has no attribute 'total_seconds'

yeah, switched from using a `timedelta` to a float for seconds, and this is fallout from that.

EDIT: Oh now I know what the issue is. It was in-place modifications I had done

```diff
diff --git a/awx/settings/defaults.py b/awx/settings/defaults.py
index 43cbab180a..d2b242eadf 100644
--- a/awx/settings/defaults.py
+++ b/awx/settings/defaults.py
@@ -456,7 +456,7 @@ CELERYBEAT_SCHEDULE = {
 DISPATCHER_SCHEDULE = {}
 for options in CELERYBEAT_SCHEDULE.values():
     task_name = options['task']
-    DISPATCHER_SCHEDULE[task_name] = options
+    DISPATCHER_SCHEDULE[task_name] = options.copy()
     DISPATCHER_SCHEDULE[task_name]['schedule'] = options['schedule'].total_seconds()
 
 # Django Caching Configuration

```

AlanCoding commented on awx/main/tasks/system.py:
```diff
@@ -597,11 +596,127 @@ def inspect_execution_and_hop_nodes(instance_list):
+
+    # Register the alternative implementation
+    ALTERNATIVE_TASK_IMPLEMENTATIONS[cluster_node_heartbeat.name] = adispatch_cluster_node_heartbeat
+    logger.info(f"Successfully registered dispatcherd method for {cluster_node_heartbeat.name}")
```
This log gets noisy due to (what I think is) uWSGI cycling workers. It's fine for now. As we get closer, we might remove it after it's tested.

AlanCoding commented on awx/main/dispatch/publish.py:
```diff
@@ -11,6 +11,9 @@
 logger = logging.getLogger('awx.main.dispatch')
 
+# Registry of tasks with alternative implementations for the new dispatcher
+ALTERNATIVE_TASK_IMPLEMENTATIONS = {}
```
:+1: good solution

art-tapin replied:
Thanks for the approval on the registry approach,it gives us a nice central place to manage alternative implementations

AlanCoding reviewed (approved):
With the 1 modification for scheduled jobs, able to run the Demo Job Template and schedules run.

art-tapin committed:
fix(dispatcher): Fix shallow copy in dispatcher schedule conversion

This resolves "AttributeError: 'float' object has no attribute 'total_seconds'"
errors when the dispatcher is restarted.

Refs: AAP-41775

art-tapin commented:
@AlanCoding I've added the .copy() method to create a proper copy in this PR, if you're okay with it

> > AttributeError: 'float' object has no attribute 'total_seconds'
> 
> yeah, switched from using a `timedelta` to a float for seconds, and this is fallout from that.
> 
> EDIT: Oh now I know what the issue is. It was in-place modifications I had done
> 
> ```diff
> diff --git a/awx/settings/defaults.py b/awx/settings/defaults.py
> index 43cbab180a..d2b242eadf 100644
> --- a/awx/settings/defaults.py
> +++ b/awx/settings/defaults.py
> @@ -456,7 +456,7 @@ CELERYBEAT_SCHEDULE = {
>  DISPATCHER_SCHEDULE = {}
>  for options in CELERYBEAT_SCHEDULE.values():
>      task_name = options['task']
> -    DISPATCHER_SCHEDULE[task_name] = options
> +    DISPATCHER_SCHEDULE[task_name] = options.copy()
>      DISPATCHER_SCHEDULE[task_name]['schedule'] = options['schedule'].total_seconds()
>  
>  # Django Caching Configuration
> ```
